src/solver/conv_asm_implicit_gemm_v4r1_dynamic.cpp

```diff
@@ -0,0 +1,589 @@
+    InBlockCopyClusterLengths_E  = other.InBlockCopyClusterLengths_E;
+    InBlockCopyClusterLengths_N1 = other.InBlockCopyClusterLengths_N1;
+    InBlockCopyClusterLengths_B  = other.InBlockCopyClusterLengths_B;
+    InBlockCopyClusterLengths_N2 = other.InBlockCopyClusterLengths_N2;
+    WeiBlockCopyClusterLengths_E = other.WeiBlockCopyClusterLengths_E;
+    WeiBlockCopyClusterLengths_K = other.WeiBlockCopyClusterLengths_K;
+}
+
+bool PerformanceImplicitGemmV4R1Dynamic::SetNextValue()
+{
+    do
+    {
+        size_t total_kernels = GetImplicitGemmV4R1DynamicTunables().size();
+        PreGeneratedKernelIndex++;
+        if(PreGeneratedKernelIndex >= total_kernels)
+            return false;
+        Copy(GetImplicitGemmV4R1DynamicTunables()[PreGeneratedKernelIndex]);
+    } while(false);
+
+    return true;
+}
+
+std::string PerformanceImplicitGemmV4R1Dynamic::ToString() const
+{
+    std::ostringstream ss;
+    Serialize(ss);
+    return ss.str();
+}
+
+PerformanceImplicitGemmV4R1Dynamic::PerformanceImplicitGemmV4R1Dynamic(bool)
+    : PerformanceImplicitGemmV4R1Dynamic(
+          -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1)
+{
+    // get the minimal routine
+    Copy(GetImplicitGemmV4R1DynamicTunables()[0]);
+    PreGeneratedKernelIndex = 0;
+}
+
+PerformanceImplicitGemmV4R1Dynamic::PerformanceImplicitGemmV4R1Dynamic(
+    int BPerBlock_,
+    int KPerBlock_,
+    int EPerBlock_,
+    int GemmNRepeat_,
+    int GemmMPerThreadSubC_,
+    int GemmNPerThreadSubC_,
+    int GemmMLevel0Cluster_,
+    int GemmNLevel0Cluster_,
+    int GemmMLevel1Cluster_,
+    int GemmNLevel1Cluster_,
+    int InBlockCopyClusterLengths_E_,
+    int InBlockCopyClusterLengths_N1_,
+    int InBlockCopyClusterLengths_B_,
+    int InBlockCopyClusterLengths_N2_,
+    int WeiBlockCopyClusterLengths_E_,
+    int WeiBlockCopyClusterLengths_K_)
+    : BPerBlock(BPerBlock_),
+      KPerBlock(KPerBlock_),
+      EPerBlock(EPerBlock_),
+      GemmNRepeat(GemmNRepeat_),
+      GemmMPerThreadSubC(GemmMPerThreadSubC_),
+      GemmNPerThreadSubC(GemmNPerThreadSubC_),
+      GemmMLevel0Cluster(GemmMLevel0Cluster_),
+      GemmNLevel0Cluster(GemmNLevel0Cluster_),
+      GemmMLevel1Cluster(GemmMLevel1Cluster_),
+      GemmNLevel1Cluster(GemmNLevel1Cluster_),
+      InBlockCopyClusterLengths_E(InBlockCopyClusterLengths_E_),
+      InBlockCopyClusterLengths_N1(InBlockCopyClusterLengths_N1_),
+      InBlockCopyClusterLengths_B(InBlockCopyClusterLengths_B_),
+      InBlockCopyClusterLengths_N2(InBlockCopyClusterLengths_N2_),
+      WeiBlockCopyClusterLengths_E(WeiBlockCopyClusterLengths_E_),
+      WeiBlockCopyClusterLengths_K(WeiBlockCopyClusterLengths_K_)
+{
+    PreGeneratedKernelIndex = 0;
+}
+
+bool ConvAsmImplicitGemmV4R1DynamicFwd::IsApplicable(const ConvolutionContext& ctx) const
+{
+    if(!ctx.direction.IsForward())
+        return false;
+
+    if(!ctx.Is2d())
+        return false;
+
+    // if(!ctx.IsFp32() && !ctx.IsFp16() && !ctx.IsBfp16())
```

Comment:
Are these Solvers applicable for all gfx architectures?

Reply:
updated a arch version check, and only applicable for gfx900, gfx906